### PR TITLE
[niconico] Add playlist extractors and refactor

### PR DESCRIPTION
**Pull request by Lesmiscore**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

This adds support for watch history and series, as well as reducing duplicated code.

### CODE REVIEW

File: yt_dlp/extractor/niconico.py [hunk 1]
```diff
@@ -606,8 +609,63 @@ def get_video_info_xml(items):
...
+
+    def _fetch_page(self, list_id, page):
+        page += 1
+        items = self._call_api(list_id, 'page %d' % page, {
+            'page': page,
+            'pageSize': self._PAGE_SIZE,
+        })['items']
+        for video in items:
+            # this is needed to support both mylist and user
+            video = traverse_obj(video, ('video',), (), expected_type=dict)
```
Comment (Lesmiscore) [inline yt_dlp/extractor/niconico.py:643]: can't it be rewritten with one `traverse_obj` invocation?

Comment (pukkandan) [reply yt_dlp/extractor/niconico.py:644]: try `traverse_obj(items, (..., ('video', None)), expected_type=dict)`

Comment (Lesmiscore) [reply yt_dlp/extractor/niconico.py:644]: The last path should behave "get `video` if there is, if not, return the object itself", but yours isn't
In brief python code, it looks like
```python
items = ...['items']
for item in items:
  video = item.get('video') or item
  ...
```

Comment (pukkandan) [reply yt_dlp/extractor/niconico.py:644]: You are right. it cant be done without a for
```
>>> from yt_dlp.utils import traverse_obj
>>> items = [{'video': 'x'}, 'y']
>>> traverse_obj(items, (..., ('video', None)))
['x', {'video': 'x'}, 'y']
>>> [traverse_obj(i, 'video', None) for i in items]
['x', 'y']
```

Comment (Lesmiscore) [reply yt_dlp/extractor/niconico.py]: but there's no problem since it checks for `id` for each items

File: yt_dlp/extractor/niconico.py [hunk 2]
```diff
@@ -606,8 +610,64 @@ def get_video_info_xml(items):
...
+
+    @staticmethod
+    def _parse_owner(item):
+        owner = item.get('owner') or {}
+        if owner:
+            return {
+                'uploader': owner.get('name'),
+                'uploader_id': owner.get('id'),
+            }
+        return {}
```
Comment (Ashish0804) [inline yt_dlp/extractor/niconico.py:634]: Isn't this better?
```suggestion
        if owner and (owner.get('id') or owner.get('name')):
            return {
                'uploader': owner.get('name'),
                'uploader_id': owner.get('id'),
            }
        return {}
```

File: yt_dlp/extractor/niconico.py [hunk 3]
```diff
@@ -606,8 +610,64 @@ def get_video_info_xml(items):
...
+                'id': video_id,
+                'title': video.get('title'),
+                'url': 'https://www.nicovideo.jp/watch/' + video_id,
+                'description': video.get('shortDescription'),
+                'duration': int_or_none(video.get('duration')),
+                'view_count': get_count('view'),
+                'comment_count': get_count('comment'),
+                'thumbnail': traverse_obj(video, ('thumbnail', ('nHdUrl', 'largeUrl', 'listingUrl', 'url'))),
+                'ie_key': NiconicoIE.ie_key(),
+                **self._parse_owner(video),
```
Comment (Ashish0804) [inline yt_dlp/extractor/niconico.py:661]: is this function really needed?
Like it's just 2 lines...that too non fatal ones.
Why not just use traverse_obj directly here?

File: yt_dlp/extractor/niconico.py [hunk 4]
```diff
@@ -618,48 +678,116 @@ class NiconicoPlaylistIE(InfoExtractor):
...
+
+        title = self._search_regex(
+            (r'<title>「(.+)（全',
+             r'<div class="TwitterShareButton"\s+data-text="(.+)\s+https:'),
+            webpage, 'title', fatal=False)
+        if title:
+            title = unescapeHTML(title)
+        playlist = []
+        for match in re.finditer(r'href="/watch/([a-z0-9]+)" data-href="/watch/\1', webpage):
+            playlist.append(self.url_result('https://www.nicovideo.jp/watch/%s' % match.group(1), video_id=match.group(1)))
```
Comment (Ashish0804) [inline yt_dlp/extractor/niconico.py:743]:
```suggestion
        for match in re.finditer(r'href="/watch/([a-z0-9]+)" data-href="/watch/\1', webpage):
            v_id = match.group(1)
            if not v_id:
                continue
            playlist.append(self.url_result(f'https://www.nicovideo.jp/watch/{v_id}', video_id=v_id))
```

File: yt_dlp/extractor/niconico.py (hunk 3 above)
Comment (Ashish0804) [reply yt_dlp/extractor/niconico.py:661]: ok nvm,
I see that this is being used at other places as well.

**Comment by Ashish0804** [comment]
1 minor nitpick, prefer to use fstrings.

File: yt_dlp/extractor/niconico.py [hunk 5]
```diff
@@ -618,48 +678,119 @@ class NiconicoPlaylistIE(InfoExtractor):
...
     def _real_extract(self, url):
         list_id = self._match_id(url)
+        mylist = self._call_api(list_id, 'list', {
+            'pageSize': 1,
+        })
+        entries = self._entries(functools.partial(self._fetch_page, list_id))
+        result = self.playlist_result(
+            entries, list_id, mylist.get('name'), mylist.get('description'))
+        result.update(self._parse_owner(mylist))
+        return result
```
Comment (pukkandan) [inline yt_dlp/extractor/niconico.py:705]:
```suggestion
        return self.playlist_result(
            entries, list_id, mylist.get('name'), mylist.get('description'), **self._parse_owner(mylist))
```

File: yt_dlp/extractor/niconico.py [hunk 6]
```diff
@@ -618,48 +678,119 @@ class NiconicoPlaylistIE(InfoExtractor):
...
+            if not v_id:
+                continue
+            playlist.append(self.url_result(f'https://www.nicovideo.jp/watch/{v_id}', video_id=v_id))
+        return self.playlist_result(playlist, list_id, title)
+
+
+class NiconicoHistoryIE(NiconicoPlaylistBaseIE):
+    IE_NAME = 'niconico:history'
+    IE_DESC = 'NicoNico user history. Requires cookies.'
+    # actual url of history page is "https://www.nicovideo.jp/my/history/video", but /video is omitted to widen matches
```
Comment (pukkandan) [inline yt_dlp/extractor/niconico.py:753]: unnecessary comment.

File: yt_dlp/extractor/niconico.py [hunk 7]
```diff
@@ -618,48 +678,119 @@ class NiconicoPlaylistIE(InfoExtractor):
...
+                'pageSize': 1,
+            })
+        except ExtractorError as e:
+            if isinstance(e.cause, compat_HTTPError) and e.cause.code == 401:
+                self.raise_login_required('You have to be logged in to get your watch history')
+            raise
+        entries = self._entries(functools.partial(self._fetch_page, list_id))
+        result = self.playlist_result(entries, list_id)
+        result.update(self._parse_owner(mylist))
+        return result
```
Comment (pukkandan) [inline yt_dlp/extractor/niconico.py:793]: same as above

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/niconico.py [hunk 8]
```diff
@@ -606,8 +610,64 @@ def get_video_info_xml(items):
...
+                'duration': int_or_none(video.get('duration')),
+                'view_count': get_count('view'),
+                'comment_count': get_count('comment'),
+                'thumbnail': traverse_obj(video, ('thumbnail', ('nHdUrl', 'largeUrl', 'listingUrl', 'url'))),
+                'ie_key': NiconicoIE.ie_key(),
+                **self._parse_owner(video),
+            }
+
+    def _entries(self, pagefunc):
+        return OnDemandPagedList(pagefunc, self._PAGE_SIZE)
```
[the remark this answers is not in the document]
Comment (Lesmiscore) [reply yt_dlp/extractor/niconico.py:665]: There were workaround for when PagedList was malfuntioning...

File: yt_dlp/extractor/niconico.py (hunk 5 above)
Comment (Lesmiscore) [reply yt_dlp/extractor/niconico.py:705]: TIL it works

**Pull request merged**